Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,619 @@
+            }
+
+            // If an exception occurred during authentication, rethrow it after logging authentication attempt
+            if ((object)authenticationException != null)
+            {
+                LastException = authenticationException;
+                LogError(authenticationException.Source, authenticationException.ToString());
+                throw authenticationException;
+            }
+
+            return IsUserAuthenticated;
+        }
+
+        
+
+        /// <summary>
+        /// Logs user authentication attempt.
+        /// </summary>
+        /// <param name="loginSuccess">true if user authentication was successful, otherwise false.</param>
+        protected virtual void LogAuthenticationAttempt(bool loginSuccess)
+        {
+          
+            if ((object)UserData != null && !string.IsNullOrWhiteSpace(UserData.Username))
+            {
+                string message = $"User \"{UserData.Username}\" login attempt {(loginSuccess ? "succeeded using OpenID Connect" : "failed")}.";
+                EventLogEntryType entryType = loginSuccess ? EventLogEntryType.SuccessAudit : EventLogEntryType.FailureAudit;
+
+                // Suffix authentication failure reason on failed logins if available
+                if (!loginSuccess && !string.IsNullOrWhiteSpace(AuthenticationFailureReason))
+                    message = string.Concat(message, " ", AuthenticationFailureReason);
+
+                // Attempt to write success or failure to the event log
+                try
+                {
+                    Log.Publish(MessageLevel.Info, MessageFlags.SecurityMessage, "AuthenticationAttempt", message);
+                    LogEvent(ApplicationName, message, entryType, 1);
+                }
+                catch (Exception ex)
+                {
+                    LogError(ex.Source, ex.ToString());
+                }
+
+                // Attempt to write success or failure to the database - we allow caller to catch any possible exceptions here so that
+                // database exceptions can be tracked separately (via LastException property) from other login exceptions, e.g., when
+                // a read-only database is being used or current user only has read-only access to database.
+                if (!string.IsNullOrWhiteSpace(SettingsCategory) && UseDatabaseLogging)
+                {
+                    AdoDataConnection database = new AdoDataConnection(SettingsCategory);
+
+                    using (IDbConnection connection = database.Connection)
+                    {
+                        connection.ExecuteNonQuery(database.ParameterizedQueryString("INSERT INTO AccessLog (UserName, AccessGranted) VALUES ({0}, {1})", "userName", "accessGranted"), UserData.Username, loginSuccess ? 1 : 0);
+                    }
```

Comment:
Does this also need to be changed in the `AdoSecurityProvider`? I left these function as they were over there.

Reply:
I see, so this came from `AdoSecurityProvider`. I suppose that code was probably written before we created the `AdoDataConnection.ExecuteNonQuery()` function.

Technically, there are no errors in the original code. The main differences are:

* In the old code, database parameters are named `@userName` and `@accessGranted` in the underlying SQL statement. In the updated code, they are named `@p0` and `@p1`.
* In the old code, `AdoDataConnection.Dispose()` never gets called. This likely generates a warning from code analysis, but the underlying connection is probably the only disposable resource used by that class.
* The old code explicitly converts the `loginSuccess` Boolean to an integer. The updated code relies on the underlying logic in `AdoDataConnection` which will only convert to an integer if the underlying database engine is Oracle or PostgreSQL.
* The updated code looks better, and provides a better example to follow.

So there's no particular need to update the code in `AdoSecurityProvider`, but it would be nice to clean things up.